### PR TITLE
PLT-7692: Create flat version of createTransactionForContract

### PR DESCRIPTION
**Pull request by bjornkihlberg**

### CODE REVIEW

File: packages/runtime/client/rest/src/index.ts [hunk 1]
```diff
@@ -100,7 +104,15 @@ export interface RestAPI {
...
   ): Promise<Transactions.GetTransactionsForContractResponse>;
-  //   createTransactionForContract: Transactions.POST; // - https://docs.marlowe.iohk.io/api/create-transactions // TODO: Jamie, lets unify names
+  /**
+   * Create an unsigned transaction which applies inputs to a contract.
+   * @see {@link https://docs.marlowe.iohk.io/api/apply-inputs-to-contract}
+   */
+  createTransactionForContract(
+    contractId: ContractId,
+    postTransactionsRequest: Transactions.PostTransactionsRequest,
+    addressesAndCollaterals: AddressesAndCollaterals
```
Comment (hrajchert) [inline packages/runtime/client/rest/src/index.ts:114]: To be consistent with other endpoints, I think this should receive a single object with clear markings on optional and required fields. Use [this](https://github.com/input-output-hk/marlowe-ts-sdk/blob/1af1b33d8a0b2047e80563e077e288f6511ae898/packages/runtime/client/rest/src/contract/endpoints/collection.ts#L203) as a starting point, and notice [the changes](https://github.com/input-output-hk/marlowe-ts-sdk/pull/42/files#diff-a61d1081ab5a7fac0cc6b84a9e54a67226c806c18fe58e2f308f40972e9bcaedR213-R239) here from my last unmerged PR that is being reviewed

For example, in addressAndCollateral only the collateral is mandatory, usedAddresses and collateralUTxOs arent.

Also, the response should probably be wrapped in its own Interface, notice to improve legibility.

Notice the exported documentation

Screenshot 2023-10-18 at 10 34 28

Comment (hrajchert) [reply packages/runtime/client/rest/src/index.ts:114]: Also, following the 1-1 mapping, the name of this endpoint should be `applyInputsToContract`.
I think the previous name (the one on the ticket) was from Jamies straw suggestion

**Review by hrajchert — CHANGES_REQUESTED** [review]
Besides the comments in the commit, this is missing the button for testing in the rest-client-flow.html

I'd wait for [this PR](https://github.com/input-output-hk/marlowe-ts-sdk/pull/42) as it simplifies its development

File: packages/runtime/client/rest/src/index.ts (hunk 1 above)
Comment (bjornkihlberg) [reply packages/runtime/client/rest/src/index.ts:114]: Alright, no problems. I saw that there was a discrepancy between our names and the names in the [Runtime API](https://docs.marlowe.iohk.io/api). I will fix this.

File: packages/runtime/client/rest/src/index.ts [hunk 2]
```diff
@@ -100,7 +110,22 @@ export interface RestAPI {
...
+    changeAddress: AddressBech32;
+    usedAddresses?: AddressBech32[];
+    collateralUTxOs?: TxOutRef[];
+    invalidBefore?: string;
+    invalidHereafter?: string;
+    version?: "v1";
+    metadata?: { [label: string | number]: any };
+    tags?: { [tag: string]: any };
+    inputs: Input[];
+  }): Promise<Transactions.TransactionTextEnvelope>;
```
Comment (bjornkihlberg) [inline packages/runtime/client/rest/src/index.ts:128]: @hrajchert Having the API consumer in mind I designed this interface. I'm curious what you think.

I think this is quite straight forward for someone invoking this. Only the `inputs: Input[]` field could arguably also be redesigned to be less obfuscated but I think that is a bigger project (but worth considering).

Comment (hrajchert) [reply packages/runtime/client/rest/src/index.ts:128]: To be consistent with other endpoints I would put it as its own type, probably named `ApplyInputsToContractRequest` or `ApplyInputsRequest`, and change `kwargs` (not sure if it stands for keyword arguments 🤔 ) for `request`

Comment (bjornkihlberg) [reply packages/runtime/client/rest/src/index.ts:128]: Alright, no worries. I'll fix it. Thanks!

Yes, `kwargs` stands for "keyword arguments". I don't mind using `request`.

File: packages/runtime/client/rest/src/contract/transaction/endpoints/collection.ts [hunk 3]
```diff
@@ -124,6 +127,19 @@ export const TransactionTextEnvelope = t.type({
...
+export type ApplyInputsToContractRequest = {
+  contractId: ContractId;
+  changeAddress: AddressBech32;
+  usedAddresses?: AddressBech32[];
+  collateralUTxOs?: TxOutRef[];
+  invalidBefore?: string;
+  invalidHereafter?: string;
+  version?: "v1";
+  metadata?: { [label: string | number]: any };
+  tags?: { [tag: string]: any };
```
Comment (nhenin) [inline packages/runtime/client/rest/src/contract/transaction/endpoints/collection.ts:139]: There is a type `Tags` I think somewhere, unless it has been removed...

File: packages/runtime/client/rest/src/contract/transaction/endpoints/collection.ts [hunk 4]
```diff
@@ -124,6 +127,19 @@ export const TransactionTextEnvelope = t.type({
...
 
+export type ApplyInputsToContractRequest = {
+  contractId: ContractId;
+  changeAddress: AddressBech32;
+  usedAddresses?: AddressBech32[];
+  collateralUTxOs?: TxOutRef[];
+  invalidBefore?: string;
+  invalidHereafter?: string;
+  version?: "v1";
+  metadata?: { [label: string | number]: any };
```
Comment (nhenin) [inline packages/runtime/client/rest/src/contract/transaction/endpoints/collection.ts:138]: the same for Metadata

File: packages/runtime/client/rest/src/contract/transaction/endpoints/collection.ts [hunk 5]
```diff
@@ -124,6 +127,19 @@ export const TransactionTextEnvelope = t.type({
   tx: TextEnvelopeGuard,
 });
 
+export type ApplyInputsToContractRequest = {
+  contractId: ContractId;
+  changeAddress: AddressBech32;
+  usedAddresses?: AddressBech32[];
+  collateralUTxOs?: TxOutRef[];
+  invalidBefore?: string;
```
Comment (nhenin) [inline packages/runtime/client/rest/src/contract/transaction/endpoints/collection.ts:135]: there are type for these dates as well

File: packages/runtime/client/rest/src/contract/transaction/endpoints/collection.ts (hunk 3 above)
Comment (nhenin) [reply packages/runtime/client/rest/src/contract/transaction/endpoints/collection.ts:139]: This is the interface in the runtime-lifecycle, it should be a superset of  it
```
export type ApplyInputsRequest = {
  inputs: Input[];
  tags?: Tags;
  metadata?: Metadata;
  invalidBefore?: ISO8601;
  invalidHereafter?: ISO8601;
};
```

Comment (nhenin) [reply packages/runtime/client/rest/src/contract/transaction/endpoints/collection.ts:139]: Same name etc..

File: packages/runtime/client/rest/src/index.ts [hunk 6]
```diff
@@ -105,7 +105,13 @@ export interface RestAPI {
     contractId: ContractId,
     range?: TransactionsRange
   ): Promise<Transactions.GetTransactionsForContractResponse>;
-  //   createTransactionForContract: Transactions.POST; // - https://docs.marlowe.iohk.io/api/create-transactions // TODO: Jamie, lets unify names
+  /**
+   * Create an unsigned transaction which applies inputs to a contract.
+   * @see {@link https://docs.marlowe.iohk.io/api/apply-inputs-to-contract}
+   */
+  applyInputsToContract(
```
Comment (nhenin) [inline packages/runtime/client/rest/src/index.ts:112]: `applyInputsBuildUnsignedTx` what do you think @hrajchert  ?

I have this in the ts-sdk : 
````
  /**
   * Submit to the Cardano Ledger, the Transaction(Tx) that will apply inputs to a given created contract.
   * @param contractId Contract Id where inputs will be applied
   * @param applyInputsRequest inputs to apply
   * @throws DecodingError
   */
  applyInputs(
    contractId: ContractId,
    applyInputsRequest: ApplyInputsRequest
  ): Promise<TxId>;
```

**Review by nhenin — DISMISSED** [review]

File: packages/runtime/client/rest/src/index.ts (hunk 2 above)
Comment (bjornkihlberg) [reply packages/runtime/client/rest/src/index.ts]: Fixed!

File: packages/runtime/client/rest/src/index.ts (hunk 6 above)
Comment (bjornkihlberg) [reply packages/runtime/client/rest/src/index.ts:125]: Currently it's mirroring the name used in the Runtime Rest API documentation ([applyInputsToContract](https://docs.marlowe.iohk.io/api/apply-inputs-to-contract)).

File: packages/runtime/client/rest/src/contract/transaction/endpoints/collection.ts [hunk 7]
```diff
@@ -129,6 +133,19 @@ export const TransactionTextEnvelope = t.type({
...
 });
 
+export type ApplyInputsToContractRequest = {
+  contractId: ContractId;
+  changeAddress: AddressBech32;
+  usedAddresses?: AddressBech32[];
+  collateralUTxOs?: TxOutRef[];
+  invalidBefore?: ISO8601;
+  invalidHereafter?: ISO8601;
+  version?: MarloweVersion;
```
Comment (hrajchert) [inline packages/runtime/client/rest/src/contract/transaction/endpoints/collection.ts:143]: something to discuss for later, do we want to add version in the API or is this something fixed for the frontend that knows that it is compatible with certain version? 🤔

**Review by hrajchert — APPROVED** [review]

**Comment by hrajchert** [comment]
This is approved but there is a a merge conflict, please fix and merge

**Pull request merged**